File: RotateArray2.cpp
/*
 * 不符合空间O(1)，空间复杂度为O(k)
 */
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        
        int n = nums.size();
        k %= n;
        
        nums.insert(nums.begin(), nums.end() - k, nums.end());
        nums.erase(nums.end() - k, nums.end());
        
    }
};

int main(){
    
    int n;
    cin>>n;
    
    vector<int> nums;
    for(int i=0;i<n;i++){
        int num;
        cin>>num;
        nums.push_back(num);
    }
    
    int k;
    cin>>k;
    
    Solution *solution = new Solution();
    solution->rotate(nums, k);
    
    for(int i=0;i<n;i++) cout<<nums[i]<<"  ";
    
    return 0;
    
}

File: RotateArray2.py
class Solution:
    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        k %= n
        
        count = 0
        start = 0
        while count < n:
            pre = nums[start]
            next = (start + k) % n
            while next != start:
                temp = nums[next]
                nums[next] = pre
                pre = temp
                next = (next + k) % n
                count += 1
            nums[start] = pre
            count += 1
            start += 1

if __name__ == '__main__':
    nums = list(input())
    k = int(raw_input())
    solution = Solution()
    solution.rotate(nums, k)
    print(nums)
